[12.0][MIG] hr_timesheet_auto_creation: migration to 12.0

huguesdk opened the pull request:
[task](https://gestion.coopiteasy.be/web#id=6770&view_type=form&model=project.task)

elicoidal committed:
Uploaded modules for migration / harmonizewd manifest / set installable false

sudhir-serpentcs committed:
[MIG][10.0-Issue175-T23293] hr_timesheet_auto_creation.

sudhir-serpentcs committed:
[IMP] README.

sudhir-serpentcs committed:
[IMP] Improved the code of UT cases to fix the travises.

vvrossem committed:
[ADD] hr_timesheet_auto_creation / backported from 10.0...

...from github : Elico-Corp 10.0/hr_timesheet_auto_creation

vvrossem committed:
[FIX] hr_timesheet_auto_creation / IntegrityError

vvrossem committed:
[ADD] hr_timesheet_auto_creation : _logger

robinkeunen committed:
[FIX] hr_timesheet_auto_creation: fix rst syntax

robinkeunen committed:
[REF] hr_timesheet_auto_creation: flake 8

huguesdk committed:
[10.0][MIG] hr_timesheet_auto_creation: migration to 10.0

huguesdk committed:
[12.0][MIG] hr_timesheet_auto_creation: migration to 12.0

huguesdk committed:
[11.0][MIG] hr_timesheet_auto_creation: migration to 11.0

huguesdk committed:
[IMP] paint it black

huguesdk committed:
[FIX] correct cron job name

huguesdk committed:
[IMP] improve __manifest__.py

* correct summary
* correct website
* remove application property (since not a full-fledged application)
* remove default installable property
* reformat

huguesdk committed:
[IMP] fix some linting errors

huguesdk committed:
[IMP] update contributors list

huguesdk commented on hr_timesheet_auto_creation/models/hr_timesheet_sheet.py:
```diff
@@ -0,0 +1,70 @@
+        return True
+
+
+class Followers(models.Model):
+    _inherit = "mail.followers"
+
+    @api.model
+    def create(self, vals):
+        """
+        Fix for IntegrityError: duplicate key value violates unique constraint
+            "mail_followers_mail_followers_res_partner_res_model_id_uniq"
+        Found on https://github.com/odoo/odoo/issues/15589#issuecomment-455635940
+        """
+        if "res_model" in vals and "res_id" in vals and "partner_id" in vals:
+            dups = self.env["mail.followers"].search(
+                [
+                    ("res_model", "=", vals.get("res_model")),
+                    ("res_id", "=", vals.get("res_id")),
+                    ("partner_id", "=", vals.get("partner_id")),
+                ]
+            )
+            if len(dups):
+                for p in dups:
+                    p.unlink()
+        res = super().create(vals)
+        return res
```
before moving this class to its own file, i would like to know if this fix is still needed in 12.0. how to reproduce the bug?

vvrossem replied:
I don't remember, but there is more info on this PR : https://github.com/coopiteasy/timesheet/pull/2

huguesdk replied:
i found the cause of the bug covered by the workaround and could thus remove the class completely.

huguesdk commented on hr_timesheet_auto_creation/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "version": "12.0.1.0.0",
+    "category": "Human Resources",
+    "author": (
+        "Coop IT Easy SCRLfs, Elico Corp, Odoo Community Association (OCA)"
```
should the oca really be there? it has been added when we backported the module from 10.0.

robinkeunen replied:
Did we backport it from OCA or Elico Corp?

huguesdk replied:
i didn’t find this module in the [timesheet oca repo](https://github.com/OCA/timesheet), so i suppose it came from elico corp directly. @vvrossem?

vvrossem replied:
Backported from [Elico-Corp](https://github.com/Elico-Corp/odoo-addons/tree/10.0/hr_timesheet_auto_creation) with the purpose to submit it to OCA (didn't happen)

huguesdk committed:
[FIX] remove hack to avoid IntegrityError

* fix cause of IntegrityError and remove hack
* modify the test to cover this

this IntegrityError was due to the fact that the same dict was re-used to
create multiple timesheets. during the creation, the provided dict is
modified, adding a default follower. this resulted in the same follower
being added multiple times (one for each loop iteration), and thus in
trying to create multiple follower records with the same values.

huguesdk committed:
[IMP] add fixme

vvrossem commented on hr_timesheet_auto_creation/models/hr_timesheet_sheet.py:
```diff
@@ -30,41 +30,17 @@ def create_employee_timesheet(self):
+                    "employee_id": employee_id,
+                    "date_start": monday,
+                    "date_end": sunday,
+                }
```
I don't understand the difference between 
`vals["employee_id"] = employee_id` and `{"employee_id": employee_id}`

huguesdk replied:
`create()` modifies the provided dict. the `create()` function of `mail.thread` contains this:
```python
# subscribe uid unless asked not to
if not self._context.get('mail_create_nosubscribe'):
    for values in vals_list:
        message_follower_ids = values.get('message_follower_ids') or []
        message_follower_ids += [(0, 0, fol_vals) for fol_vals in self.env['mail.followers']._add_default_followers(self._name, [], self.env.user.partner_id.ids, customer_ids=[])[0][0]]
        values['message_follower_ids'] = message_follower_ids
```
after the call, the provided dict contains a list of followers with `"message_follower_ids"` as key.

in the first case, it is the same dict that is reused between consecutive `create()` calls. because of this, at the next call, the same follower will be added to the list because it already exists.

in the second case, a new dict is created and used at each loop iteration, so even though the dict is modified, it works as expected.

vvrossem replied:
thank you for the cristal clear explanation :) I had no clue `create()` would modify the `dict`

huguesdk replied:
i had no clue either. it was a nasty surprise. although, when i first saw the same dict being reused between loop iterations, just changing one value, it smelled not very good. unfortunately, i did not follow that track to debug, or it would have been much quicker to find.

i don’t know whether the fact that `create()` modifies the dict is documented somewhere.
